File: chatbot_api/app/routes/auth_routes.py
from flask import Blueprint, request, jsonify
from dynamodb_models import User  # Import the DynamoDB User model
from flask_jwt_extended import create_access_token, jwt_required, get_jwt_identity
from flask_bcrypt import Bcrypt
from datetime import timedelta
import logging

logger = logging.getLogger(__name__)
auth_bp = Blueprint('auth', __name__)
bcrypt = Bcrypt()

@auth_bp.route('/api/auth/signup', methods=['POST'])
def signup():
    logger.info("Received signup request")
    data = request.json
    name = data.get('name')
    email = data.get('email')
    password = data.get('password')

    # Check if the email already exists in DynamoDB
    existing_user = User.get_user_by_email(email)
    if existing_user:
        logger.warning(f"Signup failed: Email {email} already exists")
        return jsonify({'error': 'Email already exists'}), 409

    # Hash the password and store the user in DynamoDB
    hashed_password = bcrypt.generate_password_hash(password).decode('utf-8')
    user_id = User.create_user(name=name, email=email, password=hashed_password)

    logger.info(f"User {email} signed up successfully with ID {user_id}")

    # Generate a JWT token
    expires = timedelta(days=15)
    access_token = create_access_token(identity={'email': email}, expires_delta=expires)
    return jsonify({'token': access_token}), 201


@auth_bp.route('/api/auth/login', methods=['POST'])
def login():
    logger.info("Received login request")
    data = request.json
    email = data.get('email')
    password = data.get('password')

    # Retrieve the user by email from DynamoDB
    user = User.get_user_by_email(email)
    if user and bcrypt.check_password_hash(user['password'], password):
        logger.info(f"User {email} logged in successfully")

        # Generate a JWT token
        expires = timedelta(days=15)
        access_token = create_access_token(identity={'email': user['email']}, expires_delta=expires)
        return jsonify({'token': access_token}), 200

    logger.warning(f"Login failed for email {email}")
    return jsonify({'error': 'Invalid credentials'}), 401


@auth_bp.route('/api/auth/user', methods=['GET'])
@jwt_required()
def get_current_user():
    try:
        logger.info("Fetching current user information")

        # Get the current user's identity (email) from the JWT token
        current_user = get_jwt_identity()

        # Retrieve the user by email from DynamoDB
        user = User.get_user_by_email(current_user['email'])

        if not user:
            # Return an error if the user is not found
            logger.warning(f"User not found: {current_user['email']}")
            return jsonify({'error': 'User not found'}), 404

        # Return the user's information as JSON
        user_info = {
            'id': user['id'],
            'name': user['name'],
            'email': user['email']
        }

        logger.info(f"User information fetched: {user_info}")
        return jsonify({'user': user_info}), 200

    except Exception as e:
        # Catch any exception and return an error response
        logger.error(f"Error fetching user information: {str(e)}")
        return jsonify({'error': 'An error occurred while fetching user information'}), 500

File: node_backend/routes/authRoutes.js
import express from 'express';
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { User } from '../models/dynamodbModels.js';
import getSecrets from '../utils/getSecrets.js';

const authRouter = express.Router();
const secrets = await getSecrets();

const JWT_SECRET_KEY = secrets.JWT_SECRET_KEY;

// Signup Route
authRouter.post('/signup', async (req, res) => {
    try {
        console.log("Received signup request");
        const { name, email, password } = req.body;

        // Check if the email already exists in the database
        const existingUser = await User.getUserByEmail(email);
        if (existingUser) {
            console.warn(`Signup failed: Email ${email} already exists`);
            return res.status(409).json({ error: 'Email already exists' });
        }

        // Hash the password
        const hashedPassword = await bcrypt.hash(password, 10);

        // Create a new user in the database
        const userId = await User.createUser(name, email, hashedPassword);

        console.log(`User ${email} signed up successfully with ID ${userId}`);

        // Generate a JWT token
        const accessToken = jwt.sign({ email }, JWT_SECRET_KEY, { expiresIn: '15d' });
        res.status(201).json({ token: accessToken });
    } catch (error) {
        console.error(`Error during signup: ${error.message}`);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Login Route
authRouter.post('/login', async (req, res) => {
    try {
        console.log("Received login request");
        const { email, password } = req.body;

        if (!email || typeof email !== 'string') {
            console.error(`Invalid email value: ${email} (type: ${typeof email})`);
            return res.status(400).json({ error: 'Invalid email' });
        }

        if (!password) {
            console.error("Password is missing");
            return res.status(400).json({ error: 'Password is required' });
        }

        // Retrieve the user by email from the database
        const user = await User.getUserByEmail(email);
        if (user && await bcrypt.compare(password, user.password)) {
            console.log(`User ${email} logged in successfully`);

            // Generate a JWT token
            const accessToken = jwt.sign({ email }, JWT_SECRET_KEY, { expiresIn: '15d' });
            return res.status(200).json({ token: accessToken });
        }

        console.warn(`Login failed for email ${email}`);
        res.status(401).json({ error: 'Invalid credentials' });
    } catch (error) {
        console.error(`Error during login: ${error.message}`);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Get Current User Route
authRouter.get('/user', async (req, res) => {
    try {
        console.log("Fetching current user information");

        // Get the JWT token from the request header
        const token = req.headers.authorization?.split(' ')[1];
        if (!token) {
            return res.status(401).json({ error: 'Unauthorized' });
        }

        // Verify the JWT token
        const decoded = jwt.verify(token, JWT_SECRET_KEY);
        const currentUserEmail = decoded.email;

        // Retrieve the user by email from the database
        const user = await User.getUserByEmail(currentUserEmail);
        if (!user) {
            console.warn(`User not found: ${currentUserEmail}`);
            return res.status(404).json({ error: 'User not found' });
        }

        // Return the user's information
        const userInfo = {
            id: user.id,
            name: user.name,
            email: user.email,
        };

        console.log(`User information fetched: ${JSON.stringify(userInfo)}`);
        res.status(200).json({ user: userInfo });
    } catch (error) {
        console.error(`Error fetching user information: ${error.message}`);
        res.status(500).json({ error: 'An error occurred while fetching user information' });
    }
});

export default authRouter;
